nexus/src/app/disk.rs

```diff
@@ -359,68 +407,120 @@ impl super::Nexus {
+        let (.., authz_project) = LookupPath::new(opctx, &self.db_datastore)
             .organization_name(organization_name)
             .project_name(project_name)
             .lookup_for(authz::Action::ListChildren)
             .await?;
-        Err(self.unimplemented_todo(opctx, Unimpl::Public).await)
+
+        self.db_datastore
+            .project_list_snapshots(opctx, &authz_project, pagparams)
+            .await
     }
 
     pub async fn snapshot_fetch(
         &self,
         opctx: &OpContext,
-        _organization_name: &Name,
-        _project_name: &Name,
+        organization_name: &Name,
+        project_name: &Name,
         snapshot_name: &Name,
     ) -> LookupResult<db::model::Snapshot> {
-        let lookup_type = LookupType::ByName(snapshot_name.to_string());
-        let not_found_error =
-            lookup_type.into_not_found(ResourceType::Snapshot);
-        let unimp = Unimpl::ProtectedLookup(not_found_error);
-        Err(self.unimplemented_todo(opctx, unimp).await)
+        let (.., db_snapshot) = LookupPath::new(opctx, &self.db_datastore)
+            .organization_name(organization_name)
+            .project_name(project_name)
+            .snapshot_name(snapshot_name)
+            .fetch()
+            .await?;
+
+        Ok(db_snapshot)
     }
 
     pub async fn snapshot_fetch_by_id(
         &self,
         opctx: &OpContext,
         snapshot_id: &Uuid,
     ) -> LookupResult<db::model::Snapshot> {
-        let lookup_type = LookupType::ById(*snapshot_id);
-        let not_found_error =
-            lookup_type.into_not_found(ResourceType::Snapshot);
-        let unimp = Unimpl::ProtectedLookup(not_found_error);
-        Err(self.unimplemented_todo(opctx, unimp).await)
+        let (.., db_snapshot) = LookupPath::new(opctx, &self.db_datastore)
+            .snapshot_id(*snapshot_id)
+            .fetch()
+            .await?;
+
+        Ok(db_snapshot)
     }
 
     pub async fn project_delete_snapshot(
         self: &Arc<Self>,
         opctx: &OpContext,
-        _organization_name: &Name,
-        _project_name: &Name,
+        organization_name: &Name,
+        project_name: &Name,
         snapshot_name: &Name,
     ) -> DeleteResult {
-        let lookup_type = LookupType::ByName(snapshot_name.to_string());
-        let not_found_error =
-            lookup_type.into_not_found(ResourceType::Snapshot);
-        let unimp = Unimpl::ProtectedLookup(not_found_error);
-        Err(self.unimplemented_todo(opctx, unimp).await)
+        // TODO-correctness
+        // This also requires solving how to clean up the associated resources
+        // (on-disk snapshots, running read-only downstairs) because disks
+        // *could* still be using them (if the snapshot has not yet been turned
+        // into a regular crucible volume). It will involve some sort of
+        // reference counting for volumes, and probably means this needs to
+        // instead be a saga.
```

Comment:
Sorry yeah, that is a little confusing. "A snapshot can be deleted in any state." refers to the snapshot states creating, ready, faulted, destroyed. Simply deleting it won't affect any running instance.

The current implementation leaks on-disk snapshots and associated read-only running downstairs, yes. If I create a snapshot X out of disk A, create disk B from snapshot X, there will be three volumes in the database:

| thing      | volume record |
|------------|---------------|
| disk A     | volume 1      |
| snapshot X | volume 2      |
| disk B     | volume 3      |

Volume 2 will be a copy of volume 1 (with modifications). Volume 3 will include a copy of volume 2 as a read-only parent.

With this code, a delete of snapshot X will delete the snapshot record and the associated volume record (volume 2). It will not delete the on-disk snapshot or associated running read-only downstairs, which are (if it is attached to an instance) being used by volume 3.

Reply:
Got it - the TL;DR is that the resource being leaked is "volumes associated with snapshots". If we don't have one already, could we file an issue for this?

Reply:
Definitely, opened https://github.com/oxidecomputer/omicron/issues/1632